### PR TITLE
Add Whisper speech-to-text model support with HF conversion, timestamp-seeking, and word-level timestamps

### PR DESCRIPTION
**Pull request by dameikle**
## Summary
  - Add Whisper speech-to-text model support (encoder, decoder, HF conversion, inference pipeline)
  - Implement sequential timestamp-seeking transcription with segment and word-level timestamp output via cross-attention DTW
  - Add audio data pipeline (corpus, iterator, mel spectrogram preprocessing)

  ## Details

  **Encoder & Config:**
  - `WhisperEncoder`: Conv1d stem + learned positional embeddings + transformer layers
  - `WhisperEncoderConfig` with audio params (sample_rate, n_fft, hop_length, chunk_length, etc.)
  - `WhisperModelConfig` following existing model config patterns
  - `add_key_bias` option in `TransformerConfig` for Whisper's Q/V-only attention bias

  **HF Conversion:**
  - Full `WhisperForConditionalGeneration` weight mappings
  - Encoder/decoder layer mappings, conv stem, learned positional embeddings
  - Handles Whisper-specific config (separate encoder/decoder dims, no RoPE)

  **Inference:**
  - `AudioTranslator` with forced decoder prefix (language, task, prompt tokens)
  - Token suppression from `generation_config.json`
  - Sequential timestamp-seeking: decodes 30s windows, advances by last timestamp
  - Three output modes: plain text, segment-level JSON, word-level JSON
  - Word-level timestamps via cross-attention DTW with median filtering

  **Data Pipeline:**
  - `AudioCorpus` / `AudioCorpusIterator` for loading audio files
  - Mel spectrogram computation with cached transforms
  - `tensorify_audio` for batching

  **Infrastructure fixes (non-Whisper):**
  - Triton fused MoE graceful fallback when triton unavailable (e.g. macOS)
  - `return_attn` kwarg properly passed through decoder layer variants
  
  Refs #194

### CODE REVIEW

File: eole/encoders/whisper.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+        Returns:
+            Tuple of:
+                - Encoded output (batch, time//2, hidden_size)
+                - None (transformers don't return final state)
+        """
+        x = torch.nn.functional.gelu(self.conv1(emb))
+        x = torch.nn.functional.gelu(self.conv2(x))
+        x = x.permute(0, 2, 1)
+
+        positions = torch.arange(x.size(1), device=x.device)
```
[the remark this answers is not in the document]
Comment (dameikle) [reply eole/encoders/whisper.py]: log_mel_spectrogram always pads/truncates audio to exactly chunk_length * sample_rate samples, so after the conv stem the sequence length is always exactly max_source_positions. So this can't overflow in practice

File: eole/config/models.py [hunk 2]
```diff
@@ -377,12 +387,39 @@ def _validate_transformer_decoder_config(self):
...
+    # Audio signal processing
+    sample_rate: int = Field(default=16000, description="Audio sample rate in Hz.")
+    chunk_length: int = Field(default=30, description="Audio chunk length in seconds.")
+    n_fft: int = Field(default=400, description="FFT window size for mel spectrogram.")
+    hop_length: int = Field(default=160, description="Hop length for mel spectrogram.")
+    timestamp_resolution: float = Field(default=0.02, description="Time resolution per timestamp token in seconds.")
+
+    @property
+    def data_category(self) -> str:
+        return "audio"
```
Comment (vince62s) [inline eole/config/models.py]: cf above

File: eole/inputters/text_corpus.py [hunk 3]
```diff
@@ -230,40 +236,54 @@ def __str__(self):
...
+                elif data_type == "image":
                     corpora_dict[corpus_id] = ImageTextCorpus(
                         corpus_id,
                         corpus_dict.path_txt,
                         is_train=True,
                     )
     elif task == CorpusTask.VALID:
         if CorpusName.VALID in config.data.keys():
-            if config.data_type == "text":
+            if data_type == "audio":
```
Comment (vince62s) [inline eole/inputters/text_corpus.py]: same, put audio last to make things consistent

File: eole/models/model.py [hunk 4]
```diff
@@ -1115,6 +1116,72 @@ def update_dropout(self, dropout, attention_dropout):
...
+
+    def forward(self, src, tgt, src_len, with_align=False, **kwargs):
+        """Forward pass: encode mel features, decode with cross-attention."""
+        # src is mel spectrogram (batch, num_mels, time) - no embedding needed
+        enc_out, enc_final_hs = self.encoder(src)
+        self.decoder.init_state(src=None, enc_out=enc_out, enc_final_hs=enc_final_hs)
+        dec_in = tgt[:, :-1]
+        tgt_pad_mask = dec_in.eq(self.tgt_pad_idx).unsqueeze(1)
+
+        # Audio encoder output has no padding — all-False mask
```
Comment (vince62s) [inline eole/models/model.py]: src_pad_mask in theory is not about encoder output but encoder input, if we train/infer a batch of sequences of varying length we need padding, so not sure about how whisper handles this.

File: eole/predict/beam_search.py [hunk 5]
```diff
@@ -335,6 +343,17 @@ def advance(self, log_probs, attn):
         self.select_indices = self._batch_index.view(_B * self.beam_size)
         self.topk_ids %= vocab_size
 
+        self._sum_logprobs = self._sum_logprobs.view(-1)[self.select_indices].view(_B, self.beam_size)
+        self._n_text_tokens = self._n_text_tokens.view(-1)[self.select_indices].view(_B, self.beam_size)
+        if step > self._prefix_len:
```
Comment (vince62s) [inline eole/predict/beam_search.py]: this is going to slow down beam search a lot. I am not quite sure to understand why we can't work with the current way of cumulating log_probs

Comment (dameikle) [reply eole/predict/beam_search.py]: Yes, good call. Using topk_log_probs is a far better approach. 

The one wrinkle is that for Whisper, sum_logprobs needs to exclude the prefix tokens (e.g. <|startoftranscript|><|en|><|transcribe|><|notimestamps|>) - it's used for avg log-prob in temperature fallback, which should only reflect the actual transcription. So we can't use topk_log_probs directly, but we can snapshot it at the prefix boundary and subtract.

Testing this change and results in identical WER (4.58%) on LibriSpeech test-clean with whisper-base-eole, beam_size=5.

File: eole/config/models.py [hunk 6]
```diff
@@ -834,6 +872,57 @@ def patch_size(self):
         return self.encoder.patch_size
 
 
+class WhisperModelConfig(TransformerConfig, BaseModelConfig):
+    """Configuration for Whisper speech-to-text models."""
```
[the remark this answers is not in the document]
Comment (dameikle) [reply eole/config/models.py]: I've kept WhisperModelConfig stays as it has Whisper-specific fields (suppress_tokens, etc.) but renamed the encoder config/class to AudioEncoderConfig/AudioEncoder

File: eole/config/training.py [hunk 7]
```diff
@@ -110,6 +110,12 @@ class TrainingConfig(
     )
     freeze_encoder: bool = Field(default=False, description="Freeze parameters in encoder.")
     freeze_decoder: bool = Field(default=False, description="Freeze parameters in decoder.")
+    save_frozen_weights: bool = Field(
```
[the remark this answers is not in the document]
Comment (dameikle) [reply eole/config/training.py]: Cool, I’ll split these out and raise that one

[the remark this answers is not in the document]
Comment (vince62s) [reply eole/config/training.py]: don't forget to remove in model_saver.py as well

**Comment by dameikle** [comment]
Not at all, it's great to discuss this all. I think I see what you mean. I think this would work for beam search, though for greedy search this would likely break down with the temperature scaling as different temperature attempts would have incomparably scaled scores.
